draft-ietf-quic-tls.md

```diff
@@ -760,14 +756,12 @@ and ignoring any outstanding Initial packets.
 
 ### Discarding Handshake Keys
 
-An endpoint MUST NOT discard its handshake keys until the TLS handshake is
-confirmed ({{handshake-confirmed}}).  An endpoint SHOULD discard its handshake
-keys as soon as it has confirmed the handshake.  Most application protocols
-will send data after the handshake, resulting in acknowledgements that allow
-both endpoints to discard their handshake keys promptly.  Endpoints that do
-not have reason to send immediately after completing the handshake MAY send
-ack-eliciting frames, such as PING, which will cause the handshake to be
-confirmed when they are acknowledged.
+An endpoint MUST discard its handshake keys when the TLS handshake is confirmed
```

Comment:
A "MAY" suffices here, does it not?  Or are you concerned about having to think about Handshake keys after migration?

It could be enough to say that Handshake packets MUST NOT be sent on any path other than the one on which the connection was established.

Reply:
There are a few things that could go wrong here, if keys aren't dropped. One is sending Handshake packets after a migration, another one is that a peer could retransmit Handshake packets after the handshake is confirmed, thereby consuming the complete congestion window.
Is there any reason to not make it mandatory?

Reply:
I am concerned about Handshake packets after migration, but I think @martinthomson's text seems more appropriate as a requirement. How about saying instead that Handshake packets MUST NOT be sent after the TLS handshake is confirmed?

Reply:
+1 to "MUST NOT be sent". At the same time, I do not think we need to specify when the key is discarded. Maybe "can discard" is fine?

Reply:
I'm not opposed to making this change, but I noticed that in the previous paragraph, we specify the dropping of Initial keys:

> Thus, a client MUST discard Initial keys when it first sends a Handshake packet and a server MUST discard Initial keys when it first successfully processes a Handshake packet.  Endpoints MUST NOT send Initial packets after this point.

Why do we use a MUST there, but not for Handshake keys? To prevent packet injections, it would be sufficient to say "MUST NOT process Initial packets". Should we make it consistent for both encryption levels?

Reply:
This isn't about consistency, it is about necessity.

Reply:
By having MUST dropped (or not further processes which to me is the same), there cannot be replayed packet attacks from handshake or injected packets by partially trusted co-processing logic to setup connections. I'm not sure these are valid concerns, but I don't see any benefits in keeping the handshake either. Notably, coalesced packets with a handshake packet can immediately be dropped if MUST discard is true.

Vague memory: there is a situation where a peer might want to send on invalid packet with a valid CID in order for reflection to work with ECN / RESET or something. I just don't recall what it is, only that I'm the only one disliking this "hack". However, if this "hack" relies on handshake keys being available, it won't work when they are not available, obviously.

Reply:
I'd vote for SHOULD, because I tend to think dropping them immediately is the best option, but I'm not convinced anything terrible happens if a peer delays dropping them a bit.

Reply:
> @martinthomson I don't understand your comment? What's the difference between dropping Initial keys and not processing them any more?

I was merely pointing out that the Initial keys transition is different in that continuing to handle Initial packets leaves endpoints exposed to attack longer than is ideal.  An attacker with access to Handshake keys most likely can break the remainder of the connection.

(This isn't completely true, because an attacker might only gain leaf keys and not the secrets from the core of the TLS key schedule, but the point was that access to Handshake keys signifies a much more significant breach.)